Add age column for all resources when using 'kubect get'

hurf opened the pull request:
Fixes #8929

Dispaly human readable creation time like 'age' column do in 'kubectl
get po' for all resources.

hurf committed:
Add age column for all resources when using kubect get

Dispaly human readable creation time like age column do in `kubectl
get po` for all resources.

thockin commented:
ok to test

satnam6502 commented:
As far as I can tell kubectl get pods already reports an age?

satnam6502 commented:
Ah, RCs don't.

hurf commented:
@satnam6502 It seems I don't have the permission to access the e2e result. It says `Access not possible`. Can you paste the log where the error is located here?

jlowdermilk commented:
@hurf, the e2e run timed out on cluster initialization. I believe the reason is that the addition of an AGE column on `kubectl get nodes` output broke the `validate-cluster.sh` script [here](https://github.com/GoogleCloudPlatform/kubernetes/blob/master/cluster/validate-cluster.sh#L43).

jlowdermilk commented:
Oh and regarding e2e result, the [build log](https://storage.cloud.google.com/kubernetes-jenkins/pr-logs/81a4e145ae12806e35107d83e0ed4cdb18327363/kubernetes-pull-build-test-e2e-gce/567/build-log.txt) that k8s-bot linked to above has the relevant output.

derekwaynecarr commented:
So how many columns until get pods is unreadable again? :-)

hurf commented:
@derekwaynecarr - Fortunately this patch hasn't add any column to `get po`, maybe we will worry about  `get rc` later:-)

hurf commented:
@jlowdermilk - Thanks I'll look into it.

hurf committed:
Use -o template to validate cluster

Instead of using human-oriented output, use -o template to
validate cluster in order to avoid error caused by column change.

hurf commented:
@jlowdermilk - To avoid the error may be caused by other column changes, I modified the script to find the STATUS column number instead of changing resource printing order or assigning column number directly in the script. But I'm not good at shell script, can you review the code again? I'll be happy to have your advice.

jlowdermilk commented:
LGTM.

Bash is not my strong suit either, but your changes lgtm. The shippable failure looks like a break at head that you unluckily pulled in. Try rebasing to fix.

wojtek-t commented:
@hurf - can you please rebase your PR?

bgrant0607 commented:
@derekwaynecarr One solution to lots of columns is `-o wide`: #10262. I don't think this PR cross the threshold, though.

bgrant0607 commented:
FYI @andronat

bgrant0607 commented:
@hurf Please rebase. Your tree looks like it's in an inconsistent state:
```
# github.com/GoogleCloudPlatform/kubernetes/test/integration
_output/local/go/src/github.com/GoogleCloudPlatform/kubernetes/test/integration/service_account_test.go:257: createdPod.Spec.ServiceAccount undefined (type "github.com/GoogleCloudPlatform/kubernetes/pkg/api".PodSpec has no field or method ServiceAccount)
_output/local/go/src/github.com/GoogleCloudPlatform/kubernetes/test/integration/service_account_test.go:258: createdPod.Spec.ServiceAccount undefined (type "github.com/GoogleCloudPlatform/kubernetes/pkg/api".PodSpec has no field or method ServiceAccount)
FAIL	github.com/GoogleCloudPlatform/kubernetes/test/integration [build failed]
```
I'd like to get this into 1.0.

bgrant0607 commented:
LGTM otherwise

andronat commented:
@hurf please rebase to latest version as this #9876 PR that added extra optional label columns and changed they way printing is made. You might have to heavily refactor.

hurf commented:
Sorry everybody, I was on a trip in previous days, I'll rebase it ASAP.

hurf commented:
I didn't do anything to commit 1916018 except rebase, but e2e produced 2 different results. Also shippable reported failure. Anyone can paste the test result here for me to trace the error?

andronat commented on pkg/kubectl/resource_printer.go:
```diff
@@ -600,7 +606,7 @@ func printEndpointsList(list *api.EndpointsList, w io.Writer, withNamespace bool
 
 func printNamespace(item *api.Namespace, w io.Writer, withNamespace bool, columnLabels []string) error {
-	if _, err := fmt.Fprintf(w, "%s\t%s\t%s\n", item.Name, formatLabels(item.Labels), item.Status.Phase); err != nil {
+	if _, err := fmt.Fprintf(w, "%s\t%s\t%s\t%s\n", item.Name, formatLabels(item.Labels), item.Status.Phase, translateTimestamp(item.CreationTimestamp)); err != nil {
```
Can you please delete the last `\n`? It shouldn't be there from the beginning. Thanks

jlowdermilk commented:
@hurf, shippable failure looks like flake (test-integration.sh timeout). Kicked.

hurf commented:
@andronat - done. @jlowdermilk - Thanks! can you review it again?

bgrant0607 commented on cluster/validate-cluster.sh:
```diff
@@ -32,15 +32,24 @@ attempt=0
   "${KUBE_ROOT}/cluster/kubectl.sh" get nodes > "${MINIONS_FILE}" || true
   found=$(cat "${MINIONS_FILE}" | sed '1d' | grep -c .) || true
-  ready=$(cat "${MINIONS_FILE}" | sed '1d' | awk '{print $NF}' | grep -c '^Ready') || true
+
```
Not your fault, but this is not a good idea. Scripts should:
1. specify --api-version, and
2. use -o template

The human-oriented output format may change at any time.

cc @fabioy

hurf replied:
@bgrant0607 - I was tring to reduce the influence caused by format change. And I think your opinion is more reasonable. I'll try to fix it in your way.

jlowdermilk replied:
Good point, this should really have been using `-o template` in the first place (and would then have been unaffected by your change). The following command should be useful as a starting point:
```
kubectl get nodes -o template --template='{{range .items}}{{$cond := index .status.conditions 0}}{{$cond.type}},{{$cond.status}}:{{end}}' --api-version=v1
```

hurf replied:
@jlowdermilk - Thanks for the hint. But I'm not going to use '--template' flag in the command for it is bound with api-version. I'm planning to use only '--api-version' and dump all output, then for different version, we can define different logic. I think it is good for further extension, maybe we'll have v2 or v3 later.

bgrant0607 commented:
LGTM

Restarted Shippable

cc @deads2k @ghodss

bgrant0607 commented:
Assigned to @thockin for another LGTM, since he requested this feature

bgrant0607 commented:
Please run `gofmt -s`. hack/verify-gofmt.sh will tell you which files.

Install presubmit hooks:
https://github.com/GoogleCloudPlatform/kubernetes/blob/master/docs/devel/development.md#hooks

hurf commented:
@bgrant0607 - Sorry I don't quite catch you. Does the hook related to the shippable failure?

jlowdermilk commented:
@hurf, yes the hook would catch the current shippable failure at commit time.

hurf commented:
@jlowdermilk - Still confusing. The hook looks like a gofmt check, I didn't know it is related to shippable. Anyway I'll have a try to see what would happen.

bgrant0607 commented:
@hurf Shippable runs the checks. Sorry for the confusion.

hurf commented:
@bgrant0607 - So you mean the shippable failure for this patch is cause by gofmt check? I'll look into it.

hurf commented:
@bgrant0607 - BTW, it's inconvenient that developers don't have permission to view shippable results. Hope the problem can be solved later.

bgrant0607 commented:
Re. Shippable results: I agree. #9502

wojtek-t commented:
It seems that this PR is ready for final review:
@bgrant0607 - IIUC, you already LGTMed it in https://github.com/GoogleCloudPlatform/kubernetes/pull/10008#issuecomment-115848688
@thockin - can you please take a look?

bgrant0607 commented:
Or maybe @roberthbailey, since there are changes to validate-cluster.sh.

bgrant0607 commented:
@hurf Please squash after you get another LGTM

hurf commented:
@bgrant0607 - I think the two commits are doing different things. They're not strongly bound to each other. So better to keep them seperated. What do you think?

hurf commented:
@thockin - If you have time, can you review this pr?

roberthbailey commented on cluster/validate-cluster.sh:
```diff
@@ -24,23 +24,27 @@ KUBE_ROOT=$(dirname "${BASH_SOURCE}")/..
-  "${KUBE_ROOT}/cluster/kubectl.sh" get nodes > "${MINIONS_FILE}" || true
-  found=$(cat "${MINIONS_FILE}" | sed '1d' | grep -c .) || true
-  ready=$(cat "${MINIONS_FILE}" | sed '1d' | awk '{print $NF}' | grep -c '^Ready') || true
+  "${KUBE_ROOT}/cluster/kubectl.sh" get nodes --api-version="${API_VERSION}" -o yaml> "${MINIONS_FILE}" || true
```
Why does this still use the shell script? I thought we had been transitioning to just using kubectl directly?

hurf replied:
Oh I didn't know about the transition, can you give me some hits?

jlowdermilk replied:
I think it's fine to keep using `kubectl.sh`, as it's still used by other scripts under `cluster/`. We will transition to using kubectl directly, but find/replace isn't hard.

roberthbailey commented on cluster/validate-cluster.sh:
```diff
@@ -24,23 +24,27 @@ KUBE_ROOT=$(dirname "${BASH_SOURCE}")/..
+    # when there is a 'kind: Node' line, there is a node found.
+    found=$(cat "${MINIONS_FILE}" | grep -c 'kind: Node') || true
+    # only when status: "True" and type: Ready both exist in conditions section, the node is "ready".
+    ready=$(cat "${MINIONS_FILE}" | sed -n 'N;/status: "True".*type: Ready/p; D;' | grep -c 'True') || true
```
Have you tested this sed command on both linux and mac (we've seen a lot of cases where the mac version of sed doesn't work as well as the linux version)?

hurf replied:
Unfortunately I don't have a mac environment, so I just tested it under linux. If you or anybody else can help to test the sed command under mac environment, that would be great.

bgrant0607 replied:
Does this change need to be in this PR? If not, I recommend reverting it.

hurf replied:
I've changed the output column, so if this change is not made, it never passes the test.

jlowdermilk replied:
Why not use `kubectl get -o template` here? That's pretty much what it's designed for, extracting specific values from the api object. The following command will extract `type` and `status` from each node's status.conditions[0] and output them as a comma-separated string:
```
$ kubectl get nodes -o template --template='{{range .items}}{{with index .status.conditions 0}}{{.type}}:{{.status}},{{end}}{{end}}'
Ready:True,Ready:True,%
```

hurf replied:
The reason I didn't use `-o template` is when template format changes, it will not working. But as you said, the v1 version may last longer than the script, it should not be a concern. I'll change it to `-o template`

bgrant0607 replied:
I recommend using `-o template`, but also specify the api version with --api-version=v1.

roberthbailey commented on cluster/validate-cluster.sh:
```diff
@@ -24,23 +24,27 @@ KUBE_ROOT=$(dirname "${BASH_SOURCE}")/..
+  "${KUBE_ROOT}/cluster/kubectl.sh" get nodes --api-version="${API_VERSION}" -o yaml> "${MINIONS_FILE}" || true
+  
+  found=0;ready=0
+  if [[ $API_VERSION == v1 ]]; then
```
What do we do if the api version isn't v1? We just skip these extra checks?

hurf replied:
I think if api version isn't v1, we need to define other checks. Currently it's set to v1.

jlowdermilk replied:
I would say assume the api version is v1. We'll be supporting v1 for the next year or so, which may end up being longer than the bash deployment scripts last.

roberthbailey commented on cluster/validate-cluster.sh:
```diff
@@ -57,35 +61,35 @@ while true; do
+  fi
 
-  if ((non_success_count > 0)); then
+  if ((componentstatuses > healthy)); then
```
I think that this will never work if api version != v1.

hurf replied:
There's a value initial of `componentstatuses` and `healthy`. It will not pass the check unless these two values are handled properly for each version. Currently for v1, it works. For the coming v2 or else, we need to define a different handle process for them other than for v1.

bgrant0607 replied:
Again, I don't see why this change needs to be in this PR.

hurf replied:
This change is not necessary in this PR. Just a 'by the way' work.

jlowdermilk replied:
If you are going to add this change, I again would recommend using `-o template` to extract the values:
```
$ kubectl get componentstatuses -o template --template='{{range .items}}{{with index .conditions 0}}{{.type}}:{{.status}},{{end}}{{end}}' --api-version=v1
Healthy:True,Healthy:True,Healthy:True,
```

roberthbailey commented:
@fabioy - can you take a look at the changes to validate-cluster.sh as well?

roberthbailey commented:
Also needs a rebase.

hurf commented:
@roberthbailey - Thanks for your review. And I'll try to fix the kubectl usage problem if you can give me more clues.

roberthbailey commented:
@hurf This has been marked as 1.0-post so I'll get back to you with comments after the 1.0 release.

hurf commented:
I thought @bgrant0607 once said want it be in 1.0.

hurf commented:
Ah, just saw it, it's modified to 1.0-post yesterday.

bgrant0607 commented:
@hurf I did want it, but the window has passed.

roberthbailey commented:
@hurf can you rebase and then I'll take another look. Thanks.

bgrant0607 commented:
It was good to meet you at the launch, @hurf. Please rebase. We'd like to merge PRs that accumulated during the freeze.

hurf commented:
@roberthbailey - rebased, can you take another look of the script to help improve it?

hurf commented:
@roberthbailey - can you give some advices on the script?

hurf commented:
@andronat  @jlowdermilk @thockin @erictune Can you look at this patch? It has been there for too long.

jlowdermilk commented:
@hurf, I left you some comments. Couple things to clean up, but this should be ready to merge shortly.

jlowdermilk commented on pkg/kubectl/cmd/cmd_test.go:
```diff
@@ -24,6 +24,7 @@ import (
 	"io/ioutil"
 	"os"
 	"reflect"
+	"strings"
```
Unused import. This is what's causing current shippable failure.

jlowdermilk commented:
lgtm, thanks for your patience, @hurf
